Compute the keyTimes index correctly for discrete (values) animations

Ahmad-S792 opened the pull request:
#### 580e9612e2c69282a564a14213dff82250b4bda8
<pre>
Compute the keyTimes index correctly for discrete (values) animations

Compute the keyTimes index correctly for discrete (values) animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=250643">https://bugs.webkit.org/show_bug.cgi?id=250643</a>

Reviewed by Simon Fraser.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/be50d8baacafe25f3a96d22be3e9a96ab882df4a">https://chromium.googlesource.com/chromium/blink/+/be50d8baacafe25f3a96d22be3e9a96ab882df4a</a>

This patch is to modify to ensure that the last value is allowed to be non-one
for discrete mode animations (but not for linear and spline mode animations).

* Source/WebCore/svg/SVGAnimationElement.cpp:
(SVGAnimationElement::calaculateKeyTimesIndex): Update function to exclude &apos;discrete&apos; mode
(SVGAnimationElement::calculatePercentFromKeyPoints): Add ASSERT and move functions after Disrete mode
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-elem-18-t-drt-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-elem-14-t-drt-expected.txt: Ditto
* LayoutTests/svg/animations/animate-elem-14-t-drt-expected.txt: Ditto
* LayoutTests/svg/animations/animate-elem-18-t-drt-expected.txt: Ditto

Canonical link: <a href="https://commits.webkit.org/258939@main">https://commits.webkit.org/258939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2b7be70eb66140452d851d80cdbb27b4dc89c37

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103418 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12538 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36382 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112654 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172861 "An unexpected error occured. Recent messages:Running check-change-relevance; Failed to print configuration; Running set-build-summary") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107371 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13569 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3442 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95636 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111807 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109192 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10435 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38063 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/106925 "Build is in progress. Recent messages:Pull request doesn't have relevant changes") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25097 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79794 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5928 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26500 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6102 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3028 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/102302 "Build is in progress. Recent messages:Running worker_preparation; configuring build; Pull request doesn't have relevant changes") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12085 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/36382 "Build is in progress. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/98586 "Build is in progress. Recent messages:Running configure-build; Pull request doesn't have relevant changes") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6143 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7860 "Built successfully") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/37/builds/98586 "Build is in progress. Recent messages:Running configure-build; Pull request doesn't have relevant changes") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/SVGAnimationElement.cpp:
```diff
@@ -403,10 +404,12 @@ unsigned SVGAnimationElement::calculateKeyTimesIndex(float percent) const
+    // cases we don't need to consider the last value, since |percent| is never
+    // greater than one.
+    if (keyTimesCount && calcMode() != CalcMode::Discrete)
+        keyTimesCount--;
```
We prefer `--keyTimesCount`

Ahmad-S792 replied:
Will do and then push it after last check working by api-mac. :-)

Ahmad-S792 replied:
Done and all tests pass. :-)

Ahmad-S792 committed:
Compute the keyTimes index correctly for discrete (values) animations

Compute the keyTimes index correctly for discrete (values) animations
https://bugs.webkit.org/show_bug.cgi?id=250643

Reviewed by Simon Fraser.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - https://chromium.googlesource.com/chromium/blink/+/be50d8baacafe25f3a96d22be3e9a96ab882df4a

This patch is to modify to ensure that the last value is allowed to be non-one
for discrete mode animations (but not for linear and spline mode animations).

* Source/WebCore/svg/SVGAnimationElement.cpp:
(SVGAnimationElement::calaculateKeyTimesIndex): Update function to exclude 'discrete' mode
(SVGAnimationElement::calculatePercentFromKeyPoints): Add ASSERT and move functions after Disrete mode
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-elem-18-t-drt-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/svg/animations/animate-elem-14-t-drt-expected.txt: Ditto
* LayoutTests/svg/animations/animate-elem-14-t-drt-expected.txt: Ditto
* LayoutTests/svg/animations/animate-elem-18-t-drt-expected.txt: Ditto

Canonical link: https://commits.webkit.org/258939@main
